### PR TITLE
✨ Adding uniqueness check during index process

### PR DESCRIPTION
**Pull request by policratus**
**Resolves** #67.

* Now there's a optional check, during the index process, to verify if the current image being already exists on the database;
* Although this feature is great to avoid duplicates, it creates an overhead, slowing slightly the indexing;
* This option is disabled by default.

### CODE REVIEW

File: pupyl/indexer/facets.py [hunk 1]
```diff
@@ -162,18 +162,39 @@ def append(self, tensor):
...
+
+                self.tree.build(self.size << intmul >> self.trees)
+
+                result = self.item(
+                    self.index(tensor),
+                    top=1,
+                    distances=True
+                )
+
+                if result[1][0] == 0.:
```
Comment (92amartins) [inline pupyl/indexer/facets.py:186]: Is this comparison prone to false positives? I mean, two images very similar (but not identical) to get a distance of 0? The consequence in this case would be to raise a UniqueItemError when actually that's not the case. If it is prone to false positives, how should we deal with that?

Comment (policratus) [reply pupyl/indexer/facets.py]: You're right @92amartins, it makes all the sense. Empirically testing, a difference of 0.05 should be enough to mitigate false positives (regarding similarity). Maybe on future releases, this parameter should be adaptive.

For instance, the images below have a difference of 0.053273506462574005
anitta
, being the only tiny change an artificial artefact on the neck accessory.

I updated the code so you can see and test the changes.

**Review by 92amartins — DISMISSED** [review]

**Comment by 92amartins** [comment]
@policratus, while this PR solves the problem indicated [here](https://github.com/policratus/pupyl/discussions/63), I'd like to raise a concern. Should we really raise an exception when a specific image is already index? That causes the indexing process to completely stop. Shouldn't be better to raise a warning for the user and continue the indexing process?

**Comment by policratus** [comment]
> @policratus, while this PR solves the problem indicated [here](https://github.com/policratus/pupyl/discussions/63), I'd like to raise a concern. Should we really raise an exception when a specific image is already indexed? That causes the indexing process to completely stop. Shouldn't be better to raise a warning for the user and continue the indexing process?

You're absolutely right @92amartins. I will change this behaviour to avoid unnecessary interruptions throughout the index process.

**Comment by policratus** [comment]
@92amartins, changed from raising an error to a warning:

```python
SEARCH.index('/path', check_unique=True)
pupyl/pupyl/indexer/facets.py:191: UserWarning: Tensor being indexed already exists in the database and the check for duplicates are on. Refusing to store again this tensor.
```

**Review by 92amartins — APPROVED** [review]

**Pull request merged**